pkg/workflow/argo.go

```diff
@@ -209,6 +275,63 @@ func (a *argo) generateWorkflow(ctx context.Context, g *v1alpha1.ApplicationGrou
 	return nil
 }
 
+func getTaskNamesFromHelmReleases(bucket []helmopv1.HelmRelease) []string {
+	out := []string{}
+	for _, hr := range bucket {
+		out = append(out, pkg.ConvertToDNS1123(hr.GetReleaseName()))
+	}
+	return out
+}
+
+func (a *argo) generateReversalWorkflow(ctx context.Context, l logr.Logger, nodes map[string]v1alpha12.NodeStatus, g *v1alpha1.ApplicationGroup) error {
+	graph, err := Build(g.Name, nodes)
+	if err != nil {
+		l.Error(err, "failed to build the wf status DAG")
+		return fmt.Errorf("failed to build the wf status DAG : %w", err)
+	}
+
+	rev := graph.Reverse()
+
+	entry := v1alpha12.Template{
+		Name: entrypointTplName,
+		DAG: &v1alpha12.DAGTemplate{
+			Tasks: make([]v1alpha12.DAGTask, 0),
+		},
+	}
+
+	var prevbucket []helmopv1.HelmRelease
+	for _, bucket := range rev {
+		for _, hr := range bucket {
+			task := v1alpha12.DAGTask{
+				Name:     pkg.ConvertToDNS1123(hr.GetReleaseName()),
+				Template: helmReleaseReversalExecutor,
+				Arguments: v1alpha12.Arguments{
+					Parameters: []v1alpha12.Parameter{
+						{
+							Name:  helmReleaseArg,
+							Value: strToStrPtr(hrToYAML(hr)),
+						},
+					},
+				},
+				Dependencies: convertSliceToDNS1123(getTaskNamesFromHelmReleases(prevbucket)),
```

Comment:
Dependencies may need finer tunning

Reply:
What sort of finer tuning do you have in mind?

Reply:
Described in the picture below. Assuming a forward workflow, this PR adds dependencies as shown in the middle graph. Do we need to have it like in the 3rd graph, or does it not matter?
![reversal-workflow-fine-tuning](https://user-images.githubusercontent.com/47837266/116600047-b3a36200-a8dd-11eb-9811-2ba146790ab3.JPG)

Reply:
I suppose it doesnt matter on reversal since the workflow engine will trim the graph in it's reconciler before executing the nodes.